### PR TITLE
Overall code coverage computation

### PR DESCRIPTION
**Pull request by tisnik**
## Description

Overall code coverage computation.
We had a discussion with @syedriko that it would be useful to know code coverage for both unit tests and integration tests. It is logical as we don't want to cover literally everything from unit tests and in some cases it is not even possible (`ols/app/main.py` for example). On other hand we don't want to repeat the same checks in integration tests just to have nice code coverage (`config.py`). And additionally the overall code coverage might be very near 100% - and any part not covered is either dead code or our testing is not complete.

This is how the overall code coverage look like today. It is clear that some parts of `llm_loader.py` is not tested at all. And it is not a coincidence that this is most problematic part of our codebase:

```
Name                                          Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------
ols/__init__.py                                   0      0   100%
ols/app/__init__.py                               0      0   100%
ols/app/endpoints/__init__.py                     0      0   100%
ols/app/endpoints/feedback.py                     9      0   100%
ols/app/endpoints/health.py                      15      0   100%
ols/app/endpoints/ols.py                         68      0   100%
ols/app/main.py                                  18      1    94%   33
ols/app/metrics.py                                5      0   100%
ols/app/models/__init__.py                        0      0   100%
ols/app/models/config.py                        320      8    98%   315, 355, 394, 455, 489-497, 541
ols/app/models/models.py                         30      0   100%
ols/app/routers.py                                8      0   100%
ols/constants.py                                 33      0   100%
ols/src/__init__.py                               0      0   100%
ols/src/cache/__init__.py                         0      0   100%
ols/src/cache/cache.py                           22      0   100%
ols/src/cache/cache_factory.py                   15      0   100%
ols/src/cache/in_memory_cache.py                 39      1    97%   10
ols/src/cache/redis_cache.py                     34      0   100%
ols/src/llms/__init__.py                          0      0   100%
ols/src/llms/llm_loader.py                      103     36    65%   108, 119, 150, 159, 164, 174-199, 203-235, 299-335, 339-341
ols/src/query_helpers/__init__.py                 6      0   100%
ols/src/query_helpers/docs_summarizer.py         58      5    91%   27-29, 45, 77
ols/src/query_helpers/query_docs.py              19      0   100%
ols/src/query_helpers/question_validator.py      25      0   100%
ols/src/ui/__init__.py                            0      0   100%
ols/src/ui/gradio_ui.py                          48      2    96%   94, 98
ols/utils/__init__.py                             0      0   100%
ols/utils/config.py                              32      0   100%
ols/utils/logging.py                              6      0   100%
ols/utils/suid.py                                 9      0   100%
ols/utils/token_handler.py                       41      0   100%
---------------------------------------------------------------------------
TOTAL                                           963     53    94%

```

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [x] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Run the new command `make check-coverage`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -26,24 +26,29 @@ test: test-unit test-integration test-e2e ## Run all tests
...
 
 test-integration: ## Run integration tests tests
 	@echo "Running integration tests..."
 	@echo "Reports will be written to ${ARTIFACT_DIR}"
-	python -m pytest tests/integration --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_integration.json" --junit-xml="${ARTIFACT_DIR}/junit_integration.xml" --cov-fail-under=60
+	COVERAGE_FILE=.coverage.integration python -m pytest tests/integration --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_integration.json" --junit-xml="${ARTIFACT_DIR}/junit_integration.xml" --cov-fail-under=60
 	python scripts/transform_coverage_report.py "${ARTIFACT_DIR}/coverage_integration.json" "${ARTIFACT_DIR}/coverage_integration.out"
 	scripts/codecov.sh "${ARTIFACT_DIR}/coverage_integration.out"
 
+check-coverage: test-unit test-integration  ## Unit tests and integration tests overall code coverage check
```
Comment (vishnuchalla) [inline Makefile:40]: just curious if we want to add this as a check in PROW?

Comment (tisnik) [reply Makefile:40]: in theory yes, I just need to figure out how to do it as currently unit tests are run independently of integration tests

File: Makefile [hunk 2]
```diff
@@ -26,24 +26,29 @@ test: test-unit test-integration test-e2e ## Run all tests
...
+	coverage report -m --fail-under=94
+
 test-e2e: ## Run e2e tests - requires running OLS server
 	@echo "Running e2e tests..."
 	@echo "Reports will be written to ${ARTIFACT_DIR}"
 	python -m pytest tests/e2e --junit-xml="${ARTIFACT_DIR}/junit_e2e.xml"
 		
 
-coverage-report:	test-unit ## Export unit test coverage report into interactive HTML
+coverage-report:	.coverage_unit  ## Export unit test coverage report into interactive HTML
```
Comment (bparees) [inline Makefile:51]: can you explain this change?  there is no ".coverage_unit" target, nor does any step seem to create that file.

Comment (tisnik) [reply Makefile:51]: `COVERAGE_FILE=.coverage.unit python -m pytest ...` command creates it, it's `test-unit` target

Comment (tisnik) [reply Makefile:51]: hmm thinking about it more, I'll just use the original line 46 as is. It was an optimization not to run tests again when the coverage file exists. But yes, as you said - Make does not know if the file is refreshed or not (not having knowledge that every single line in the project can change it basically).

Comment (tisnik) [reply Makefile]: updated @bparees

File: Makefile [hunk 3]
```diff
@@ -26,17 +26,22 @@ test: test-unit test-integration test-e2e ## Run all tests
...
-	python -m pytest tests/unit --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_unit.json" --junit-xml="${ARTIFACT_DIR}/junit_unit.xml"
+	COVERAGE_FILE=.coverage.unit python -m pytest tests/unit --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_unit.json" --junit-xml="${ARTIFACT_DIR}/junit_unit.xml"
 	python scripts/transform_coverage_report.py "${ARTIFACT_DIR}/coverage_unit.json" "${ARTIFACT_DIR}/coverage_unit.out"
 	scripts/codecov.sh "${ARTIFACT_DIR}/coverage_unit.out"
 
 test-integration: ## Run integration tests tests
 	@echo "Running integration tests..."
 	@echo "Reports will be written to ${ARTIFACT_DIR}"
-	python -m pytest tests/integration --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_integration.json" --junit-xml="${ARTIFACT_DIR}/junit_integration.xml" --cov-fail-under=60
+	COVERAGE_FILE=.coverage.integration python -m pytest tests/integration --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_integration.json" --junit-xml="${ARTIFACT_DIR}/junit_integration.xml" --cov-fail-under=60
```
Comment (bparees) [inline Makefile:36]: can we put this in $ARTIFACT_DIR like we do everything else so we're not polluting directories?

File: Makefile [hunk 4]
```diff
@@ -26,17 +26,22 @@ test: test-unit test-integration test-e2e ## Run all tests
 test-unit: ## Run the unit tests
 	@echo "Running unit tests..."
 	@echo "Reports will be written to ${ARTIFACT_DIR}"
-	python -m pytest tests/unit --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_unit.json" --junit-xml="${ARTIFACT_DIR}/junit_unit.xml"
+	COVERAGE_FILE=.coverage.unit python -m pytest tests/unit --cov=ols --cov-report term-missing --cov-report "json:${ARTIFACT_DIR}/coverage_unit.json" --junit-xml="${ARTIFACT_DIR}/junit_unit.xml"
```
Comment (bparees) [inline Makefile:29]: can we put this in $ARTIFACT_DIR like we do everything else so we're not polluting directories?

File: Makefile (hunk 3 above)
Comment (tisnik) [reply Makefile:36]: it's possible, sure. Btw today it is created in $CWD

Comment (tisnik) [reply Makefile]: updated @bparees

File: Makefile (hunk 4 above)
Comment (tisnik) [reply Makefile]: updated @bparees

**Pull request merged**